internal/handler/common/update.go

```diff
@@ -31,29 +33,173 @@ import (
-	var docUpdated bool
-	var err error
+// UpdateDocument iterates through documents from iter and processes them sequentially based on param.
+// Returns UpdateResult if all operations (update/upsert) are successful.
+//
+// In case of updating multiple documents, UpdateDocument returns an error immediately after one of the
+// operation fails. The rest of the documents are not processed.
+// TODO https://github.com/FerretDB/FerretDB/issues/2612
+func UpdateDocument(ctx context.Context, c backends.Collection, cmd string, iter types.DocumentsIterator, param *Update) (*UpdateResult, error) { //nolint:lll // for readability
+	result := new(UpdateResult)
 
-	if update.Len() == 0 {
-		// replace to empty doc
-		for _, key := range doc.Keys() {
-			docUpdated = true
+	docs, err := iterator.ConsumeValues(iter)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	result.Matched.Count = int32(len(docs))
+
+	var upsert bool
+
+	if len(docs) == 0 {
+		if !param.Upsert {
+			return result, nil
+		}
+
+		upsert = true
+		docs = append(docs, must.NotFail(types.NewDocument()))
+	}
+
+	for _, doc := range docs {
+		var modified bool
+
+		result.Matched.Doc = doc.DeepCopy()
+
+		if upsert {
+			if err = processFilterEqualityCondition(doc, param.Filter); err != nil {
+				return nil, lazyerrors.Error(err)
+			}
+		}
+
+		if !param.HasUpdateOperators {
+			modified, err = processReplacementDoc(cmd, doc, param.Update)
+		} else {
+			modified, err = processUpdateOperator(cmd, doc, param.Update, upsert)
+		}
+
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		if !doc.Has("_id") {
+			doc.Set("_id", types.NewObjectID())
+		}
+
+		// TODO https://github.com/FerretDB/FerretDB/issues/3454
+		if err = doc.ValidateData(); err != nil {
+			return nil, lazyerrors.Error(err)
+		}
 
-			if key != "_id" {
-				doc.Remove(key)
+		if upsert {
+			_, err = c.InsertAll(ctx, &backends.InsertAllParams{Docs: []*types.Document{doc}})
+			if err != nil {
+				return nil, lazyerrors.Error(err)
+			}
+
+			result.Upserted.Doc = doc
+		} else if modified {
+			_, err := c.UpdateAll(ctx, &backends.UpdateAllParams{Docs: []*types.Document{doc}})
+			if err != nil {
+				return nil, lazyerrors.Error(err)
+			}
+
+			result.Modified.Doc = doc
+			result.Modified.Count++
+		}
+	}
+
+	return result, nil
+}
+
+// processFilterEqualityCondition copies the fields with equality condition from filter to doc.
+func processFilterEqualityCondition(doc, filter *types.Document) error {
+	iter := filter.Iterator()
+	defer iter.Close()
+
+	for {
+		key, val, err := iter.Next()
+		if err != nil {
+			if errors.Is(err, iterator.ErrIteratorDone) {
+				return nil
+			}
+
+			return lazyerrors.Error(err)
+		}
+
+		if key[0] == '$' {
+			continue
+		}
+
+		if valDoc, ok := val.(*types.Document); ok {
+			if valDoc.Len() == 1 {
+				valEq, _ := valDoc.Get("$eq")
+				if valEq == nil {
+					continue
+				}
+				val = valEq
+			} else {
+				val = valDoc
```

Comment:
Test case `UpsertQueryCompareObject` has been added to cover this.

Reply:
Actually, `else` statement could be totally removed.

Reply:
I'm not sure how this `else` case can be totally removed. I have refactored the func because of the previous comment. Let me know if there are any good alternatives.

Reply:
Thanks, yeah let's keep as it is.